### PR TITLE
feat: Add addOrUpdate, update, and remove to PathTrie

### PR DESCRIPTION
**Pull request by nielsenko**
## Description
Add missing CRUD operations to `PathTrie`:
- `addOrUpdate`
- `update`, and
- `remove`

## Related Issues
- Closes: ?

## Pre-Launch Checklist
Please ensure that your PR meets the following requirements before submitting:

- [x] This update focuses on a single feature or bug fix. (For multiple fixes, please submit separate PRs.)
- [x] I have read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code usixg [dart format](https://dart.dev/tools/dart-format).
- [x] I have referenced at least one issue this PR fixes or is related to.
- [x] I have updated/added relevant documentation (doc comments with `///`), ensuring consistency with existing project documentation.
- [x] I have added new tests to verify the changes.
- [x] All existing and new tests pass successfully.
- [x] I have documented any breaking changes below.

## Breaking Changes
- [ ] Includes breaking changes.
- [x] No breaking changes.

## Summary by CodeRabbit

- **New Features**
  - Enhanced route management with new methods to add, update, or remove routes in the routing system.

- **Documentation**
  - Improved documentation for error conditions when attaching routes.

- **Tests**
  - Added comprehensive tests for adding, updating, and removing routes, including handling of parameterized paths and various edge cases.

### CODE REVIEW

**Comment by nielsenko** [comment]
@coderabbitai review

**Review by SandPod — APPROVED** [review]
Implementation looks good, just clean up the tests a bit. I think the comment is redundant.

File: test/router/path_trie_crud_test.dart [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+    setUp(() {
+      trie = PathTrie<int>();
+    });
+
+    group('addOrUpdate', () {
+      test(
+          'Given an empty trie, '
+          'when a new path is added with addOrUpdate, '
+          'then the path is added, lookup returns the value, and addOrUpdate returns false (added)',
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply test/router/path_trie_crud_test.dart:17]: Arbitrary choice. I'll change it.

[the remark this answers is not in the document]
Comment (nielsenko) [reply test/router/path_trie_crud_test.dart:17]: Could also do java style and return the previous value?

[the remark this answers is not in the document]
Comment (nielsenko) [reply test/router/path_trie_crud_test.dart:17]: Or an enum, to avoid ambiguity?

I don't think Dart has an established pattern here.
There is `Map.update` which uses callbacks to handle both situations (pure update, vs. add or update), but I'm not super fan of that signature.

File: test/router/path_trie_crud_test.dart [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+            reason: 'Should return false as it was a new entry.');
+        expect(result, isNotNull);
+        expect(result!.value, equals(value));
+        expect(result.parameters, isEmpty);
+      });
+
+      test(
+          'Given a trie with an existing path, '
+          'when addOrUpdate is called for the same path with a new value, '
+          'then the value is updated, lookup returns the new value, and addOrUpdate returns true (updated)',
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply test/router/path_trie_crud_test.dart]: I have simplified the then part to just `the the value is updated`.

File: test/router/path_trie_crud_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (SandPod) [reply test/router/path_trie_crud_test.dart]: Yeah, not a big fan of the callbacks either. This is kind of binary, so I would expect the response to be in consideration to the `add` part.

Enum could also work, a bit more vebose but perhaps not an issue?

[the remark this answers is not in the document]
Comment (nielsenko) [reply test/router/path_trie_crud_test.dart]: I'll stick with reverting, so `true` means `add` for now. If Dart ever gets short syntax for enums (`.added` - like rust), then let us reconsider.
